app/src/main/java/fi/metatavu/muisti/exhibitionui/files/OfflineFileController.kt

```diff
@@ -0,0 +1,231 @@
+         * @param file file
+         * @return eTag or null if not found
+         */
+        private fun readFileETag(file: File): String? {
+            val fileMeta = if (file.exists()) readFileMeta(file) else null
+            return fileMeta?.eTag
+        }
+
+        /**
+         * Read file meta
+         *
+         * @return file meta
+         */
+        private fun readFileMeta(file: File): FileMeta? {
+            if (!file.exists()) {
+                return null
+            }
+
+            val metaFile = File(file.parent, "${file.name}.meta")
+            if (!metaFile.exists()) {
+                return null
+            }
+
+            val json = metaFile.bufferedReader().use {
+                it.readText()
+            }
+
+            return getFileMetaAdapter().fromJson(json)
+        }
+
+        /**
+         * Writes file meta
+         *
+         * @param file meta
+         * @param eTag eTag
+         */
+        private fun writeFileMeta(file: File, eTag: String?) {
+            val metaFile = File(file.parent, "${file.name}.meta")
+            if (metaFile.exists()) {
+                metaFile.delete()
+            }
+
+            metaFile.parentFile?.mkdirs()
+            metaFile.createNewFile()
+
+            FileOutputStream(metaFile).use {
+                it.write(getFileMetaAdapter().toJson(FileMeta(eTag = eTag)).toByteArray())
+            }
+        }
+
+        /**
+         * Returns JsonAdapter for FileMeta
+         *
+         * @return JsonAdapter for FileMeta
+         */
+        private fun getFileMetaAdapter(): JsonAdapter<FileMeta> {
+            return getMoshi().adapter(FileMeta::class.java)
+        }
+
+        /**
+         * Returns initialized Moshi instance
+         *
+         * @return initialized Moshi instance
+         */
+        private fun getMoshi(): Moshi {
+            return Moshi.Builder()
+                .add(KotlinJsonAdapterFactory())
+                .build()
+        }
+
+        /**
+         * Calculates a md5 hash from given string
+         *
+         * @param str string
+         * @return md5 hash
+         */
+        private fun md5(str: String): String {
+            val hexString = StringBuilder()
+
+            for (aMessageDigest in md5(str.toByteArray(StandardCharsets.UTF_8))) {
+                hexString.append(Integer.toHexString(0xFF and aMessageDigest.toInt()).padStart(2, '0'))
+            }
+
+            return hexString.toString()
+        }
+
+        /**
+         * Calculates a md5 hash from given bytes
+         *
+         * @param bytes bytes
+         * @return md5 hash bytes
+         */
+        private fun md5(bytes: ByteArray): ByteArray {
+            val digest = MessageDigest.getInstance("MD5")
+            digest.update(bytes)
+            return digest.digest()
+        }
+    }
+
+}
+
+/**
+ * Private class for describing file meta JSON
+ *
+ * @param eTag file eTag
+ */
+private class FileMeta(val eTag: String?) {
+
+}
```

Comment:
Use data class instead.
```suggestion
/**
 * Private class for describing file meta JSON
 *
 * @param eTag file eTag
 */
private data class FileMeta(val eTag: String?)
```